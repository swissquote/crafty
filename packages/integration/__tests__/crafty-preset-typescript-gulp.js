const test = require("ava");
const configuration = require("@swissquote/crafty/src/configuration");
const testUtils = require("../utils");

const getCrafty = configuration.getCrafty;

test("Loads crafty-preset-typescript and does not register gulp tasks", t => {
  const crafty = getCrafty(["@swissquote/crafty-preset-typescript"], {});

  const loadedPresets = crafty.config.loadedPresets.map(
    preset => preset.presetName
  );
  t.truthy(loadedPresets.includes("@swissquote/crafty-preset-typescript"));

  crafty.createTasks();
  t.deepEqual(Object.keys(crafty.undertaker._registry.tasks()), []);
});

test("Loads crafty-preset-typescript, crafty-runner-gulp and registers gulp task", t => {
  const config = { js: { myBundle: { source: "js/**/*.ts" } } };
  const crafty = getCrafty(
    ["@swissquote/crafty-preset-typescript", "@swissquote/crafty-runner-gulp"],
    config
  );

  const loadedPresets = crafty.config.loadedPresets.map(
    preset => preset.presetName
  );
  t.truthy(loadedPresets.includes("@swissquote/crafty-preset-typescript"));
  t.truthy(loadedPresets.includes("@swissquote/crafty-runner-gulp"));

  crafty.createTasks();
  t.deepEqual(Object.keys(crafty.undertaker._registry.tasks()), [
    "js_myBundle",
    "js",
    "default"
  ]);
});

test.serial("Compiles TypeScript", async t => {
  const cwd = await testUtils.getCleanFixtures(
    "crafty-preset-typescript-gulp/compiles"
  );

  const result = await testUtils.run(["run", "default"], cwd);

  t.snapshot(result);
  t.is(result.status, 0);

  t.falsy(testUtils.exists(cwd, "dist/js/myBundle.min.js"));
  t.falsy(testUtils.exists(cwd, "dist/js/myBundle.min.js.map"));

  t.truthy(testUtils.exists(cwd, "dist/js/script.js"));
  t.truthy(testUtils.exists(cwd, "dist/js/script.js.map"));

  t.truthy(testUtils.exists(cwd, "dist/js/Component.js"));
  t.truthy(testUtils.exists(cwd, "dist/js/Component.js.map"));

  t.snapshot(testUtils.readForSnapshot(cwd, "dist/js/script.js"));
  t.snapshot(testUtils.readForSnapshot(cwd, "dist/js/Component.js"));
});

test.serial("Compiles TypeScript, keeps runtime external", async t => {
  const cwd = await testUtils.getCleanFixtures(
    "crafty-preset-typescript-gulp/compiles-import-runtime"
  );

  const result = await testUtils.run(["run", "default"], cwd);

  t.snapshot(result);
  t.is(result.status, 0);

  t.falsy(testUtils.exists(cwd, "dist/js/myBundle.min.js"));
  t.falsy(testUtils.exists(cwd, "dist/js/myBundle.min.js.map"));

  t.truthy(testUtils.exists(cwd, "dist/js/script.js"));
  t.truthy(testUtils.exists(cwd, "dist/js/script.js.map"));

  t.truthy(testUtils.exists(cwd, "dist/js/Component.js"));
  t.truthy(testUtils.exists(cwd, "dist/js/Component.js.map"));

  t.snapshot(testUtils.readForSnapshot(cwd, "dist/js/script.js"));
  t.snapshot(testUtils.readForSnapshot(cwd, "dist/js/Component.js"));
});

test.serial("Compiles TypeScript and concatenates", async t => {
  const cwd = await testUtils.getCleanFixtures(
    "crafty-preset-typescript-gulp/concatenates"
  );

  const result = await testUtils.run(["run", "default"], cwd);

  t.snapshot(result);
  t.is(result.status, 0);

  t.truthy(testUtils.exists(cwd, "dist/js/myBundle.min.js"));
  t.truthy(testUtils.exists(cwd, "dist/js/myBundle.min.js.map"));

  t.falsy(testUtils.exists(cwd, "dist/js/script.js"));
  t.falsy(testUtils.exists(cwd, "dist/js/script.js.map"));

  t.falsy(testUtils.exists(cwd, "dist/js/otherfile.js"));
  t.falsy(testUtils.exists(cwd, "dist/js/otherfile.js.map"));

  t.snapshot(testUtils.readForSnapshot(cwd, "dist/js/myBundle.min.js"));
});

test.serial("Fails gracefully on broken markup", async t => {
  const cwd = await testUtils.getCleanFixtures(
    "crafty-preset-typescript-gulp/fails"
  );

  const result = await testUtils.run(["run", "default"], cwd);

  t.snapshot(result);
  t.is(result.status, 1);

  // Files aren't generated on failed lint
  t.falsy(testUtils.exists(cwd, "dist/js/myBundle.min.js"));
  t.falsy(testUtils.exists(cwd, "dist/js/myBundle.min.js.map"));
});

test.serial("Lints TypeScript", async t => {
  const cwd = await testUtils.getCleanFixtures(
    "crafty-preset-typescript-gulp/lints"
  );

  const result = await testUtils.run(["run", "default"], cwd);

  t.snapshot(result);
  t.is(result.status, 1);

  // Files aren't generated on failed lint
  t.falsy(testUtils.exists(cwd, "dist/js/myBundle.min.js"));
  t.falsy(testUtils.exists(cwd, "dist/js/myBundle.min.js.map"));
});
