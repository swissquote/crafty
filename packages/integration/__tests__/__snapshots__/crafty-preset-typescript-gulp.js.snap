// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiles TypeScript 1`] = `
Object {
  "status": 0,
  "stdall": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-typescript-gulp/compiles/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
\`resolve/lib/core\` is deprecated; please use \`is-core-module\` directly
[__:__:__] Finished 'js_myBundle' after ____ ms
[__:__:__] Finished 'js' after ____ ms
[__:__:__] Finished 'default' after ____ ms
",
}
`;

exports[`Compiles TypeScript 2`] = `
"function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError(\\"Cannot call a class as a function\\"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if (\\"value\\" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

import test from \\"./Component\\";

var NewStuff = /*#__PURE__*/function () {
  function NewStuff() {
    _classCallCheck(this, NewStuff);

    // eslint-disable-next-line no-console
    console.log(\\"init class\\");
  }

  _createClass(NewStuff, [{
    key: \\"method\\",
    value: function method() {
      // eslint-disable-next-line no-console
      console.log(test(2, 4));
      import(\\"./SomeLibrary\\").then(function (SomeLibrary) {
        var som = new SomeLibrary.default(); // eslint-disable-next-line no-console

        console.log(som.add(2, 4));
      });
    }
  }]);

  return NewStuff;
}();

export { NewStuff as default };
//# sourceMappingURL=script.js.map
"
`;

exports[`Compiles TypeScript 3`] = `
"export default function test(a, b) {
  return a + b;
}
//# sourceMappingURL=Component.js.map
"
`;

exports[`Compiles TypeScript and concatenates 1`] = `
Object {
  "status": 0,
  "stdall": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-typescript-gulp/concatenates/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
\`resolve/lib/core\` is deprecated; please use \`is-core-module\` directly
[__:__:__] Finished 'js_myBundle' after ____ ms
[__:__:__] Finished 'js' after ____ ms
[__:__:__] Finished 'default' after ____ ms
",
}
`;

exports[`Compiles TypeScript and concatenates 2`] = `
"export default function test(a, b) {
  return a + b;
}
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError(\\"Cannot call a class as a function\\"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if (\\"value\\" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var SomeLibrary = /*#__PURE__*/function () {
  function SomeLibrary() {
    _classCallCheck(this, SomeLibrary);
  }

  _createClass(SomeLibrary, [{
    key: \\"add\\",
    value: function add(a, b) {
      return a + b;
    }
  }, {
    key: \\"substract\\",
    value: function substract(a, b) {
      return a - b;
    }
  }, {
    key: \\"divide\\",
    value: function divide(a, b) {
      return a / b;
    }
  }]);

  return SomeLibrary;
}();

export { SomeLibrary as default };
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError(\\"Cannot call a class as a function\\"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if (\\"value\\" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

import test from \\"./Component\\";

var NewStuff = /*#__PURE__*/function () {
  function NewStuff() {
    _classCallCheck(this, NewStuff);

    // eslint-disable-next-line no-console
    console.log(\\"init class\\");
  }

  _createClass(NewStuff, [{
    key: \\"method\\",
    value: function method() {
      // eslint-disable-next-line no-console
      console.log(test(2, 4));
      import(\\"./SomeLibrary\\").then(function (SomeLibrary) {
        var som = new SomeLibrary.default(); // eslint-disable-next-line no-console

        console.log(som.add(2, 4));
      });
    }
  }]);

  return NewStuff;
}();

export { NewStuff as default };
//# sourceMappingURL=myBundle.min.js.map
"
`;

exports[`Compiles TypeScript, keeps runtime external 1`] = `
Object {
  "status": 0,
  "stdall": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-typescript-gulp/compiles-import-runtime/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
\`resolve/lib/core\` is deprecated; please use \`is-core-module\` directly
[__:__:__] Finished 'js_myBundle' after ____ ms
[__:__:__] Finished 'js' after ____ ms
[__:__:__] Finished 'default' after ____ ms
",
}
`;

exports[`Compiles TypeScript, keeps runtime external 2`] = `
"import _classCallCheck from \\"@babel/runtime/helpers/classCallCheck\\";
import _createClass from \\"@babel/runtime/helpers/createClass\\";
import test from \\"./Component\\";

var NewStuff = /*#__PURE__*/function () {
  function NewStuff() {
    _classCallCheck(this, NewStuff);

    // eslint-disable-next-line no-console
    console.log(\\"init class\\");
  }

  _createClass(NewStuff, [{
    key: \\"method\\",
    value: function method() {
      // eslint-disable-next-line no-console
      console.log(test(2, 4));
      import(\\"./SomeLibrary\\").then(function (SomeLibrary) {
        var som = new SomeLibrary.default(); // eslint-disable-next-line no-console

        console.log(som.add(2, 4));
      });
    }
  }]);

  return NewStuff;
}();

export { NewStuff as default };
//# sourceMappingURL=script.js.map
"
`;

exports[`Compiles TypeScript, keeps runtime external 3`] = `
"export default function test(a, b) {
  return a + b;
}
//# sourceMappingURL=Component.js.map
"
`;

exports[`Fails gracefully on broken markup 1`] = `
Object {
  "status": 1,
  "stdall": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-typescript-gulp/fails/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
\`resolve/lib/core\` is deprecated; please use \`is-core-module\` directly
[__:__:__]
__PATH__/packages/integration/fixtures/crafty-preset-typescript-gulp/fails/js/SomeLibrary.ts
  10:0  error  Parsing error: '}' expected

✖ 1 problem (1 error, 0 warnings)

[__:__:__] 'js_myBundle' errored after ____ ms
[__:__:__] ESLint failed with 1 error
[__:__:__] 'js' errored after ____ ms
[__:__:__] 'default' errored after ____ ms
js/SomeLibrary.ts(10,1): error TS1005: '}' expected.
TypeScript: 1 syntax error
TypeScript: emit succeeded (with errors)
Error: TypeScript: Compilation failed
    ...stacktrace...
",
}
`;

exports[`Lints TypeScript 1`] = `
Object {
  "status": 1,
  "stdall": "
[__:__:__] Starting Crafty __version__...
[__:__:__] Files will be stored in:
            js: __PATH__/packages/integration/fixtures/crafty-preset-typescript-gulp/lints/dist/js
[__:__:__] Starting 'default' ...
[__:__:__] Starting 'js' ...
[__:__:__] Starting 'js_myBundle' ...
\`resolve/lib/core\` is deprecated; please use \`is-core-module\` directly
[__:__:__]
__PATH__/packages/integration/fixtures/crafty-preset-typescript-gulp/lints/js/Component.ts
  1:1  error  Replace \`⏎export·function·test(a:\` with \`export·function·test(a:·\`  @swissquote/swissquote/prettier/prettier
  3:1  error  Replace \`···return·a·+···\` with \`··return·a·+\`                      @swissquote/swissquote/prettier/prettier

__PATH__/packages/integration/fixtures/crafty-preset-typescript-gulp/lints/js/SomeLibrary.ts
  10:15  error  Delete \`⏎·\`  @swissquote/swissquote/prettier/prettier

__PATH__/packages/integration/fixtures/crafty-preset-typescript-gulp/lints/js/script.ts
   1:1   error    Replace \`⏎import·{\` with \`import·{·\`                   @swissquote/swissquote/prettier/prettier
   5:1   error    Replace \`····constructor·\` with \`··constructor\`        @swissquote/swissquote/prettier/prettier
   6:5   warning  Unexpected console statement                           no-console
   7:1   error    Insert \`·\`                                             @swissquote/swissquote/prettier/prettier
  10:5   warning  Unexpected console statement                           no-console
  10:17  error    Replace \`·test(·\` with \`test(\`                         @swissquote/swissquote/prettier/prettier
  12:34  error    Replace \`·(SomeLibrary)·=>⏎···\` with \`SomeLibrary·=>\`  @swissquote/swissquote/prettier/prettier
  16:7   warning  Unexpected console statement                           no-console
  17:7   error    Insert \`;\`                                             @swissquote/swissquote/prettier/prettier

✖ 12 problems (9 errors, 3 warnings)
  9 errors and 0 warnings potentially fixable with the \`--fix\` option.

[__:__:__] 'js_myBundle' errored after ____ ms
[__:__:__] ESLint failed with 9 errors
[__:__:__] 'js' errored after ____ ms
[__:__:__] 'default' errored after ____ ms
",
}
`;
